### ISSUE TITLE
Cache algorithm fetches based on recently used algorithm names

### ISSUE THREAD

**Issue by nhorman** [openssl/project#540]
The goal here is to create a top level cache in an OSSL_LIB_CTX that allows for created algorithms to be cached and reused, avoiding repeated calls deep into the inner_evp_generic_fetch path when possible.  i.e. like we recommend in our documentation, EVP objects should be reused when possible, rather than freed and re-created.

specifically I'm proposing that we:
* Create a hash table in an OSSL_LIB_CTX object, initialized on startup, and flushed/cleared on context free
* The cache should be keyed against the <name, propquery> tuple
* For every EVP_\<type\>_new call that we have, get and query this hash table for an already existing instance of that object
   - If found, up the ref count on that object and return it
   - Else, create the object, up the ref count, and add it to the cache
* flush the cache whenever a provider is loaded/unloaded, as that will change what the propquery resolves to

**Comment by levitte** [comment]
Actually, we do cache created methods.  See `inner_evp_generic_fetch()` in crypto/evp/evp_fetch.c.  *All* fetching functionality is constructed the same way (i.e. also in crypto/store).

**Comment by nhorman** [comment]
we discussed this a number of weeks ago, this is meant to improve on that.  When inner_evp_generic_fetch is called it incurs the expense of serveral in-order hash table traversals through the ossl_namemap, which slows things down quite alot.  This is meant to cache the completed objects at the top of these function entry points with a single hashed lookup.  I made a first draft implementation here:
https://github.com/nhorman/openssl/commit/8c1d70ebff1cc750aaa7a1773229f46ecc878772

Early Performance numbers are available here:
https://github.com/openssl/project/issues/440#issuecomment-1929357718
